crates/node/src/lib.rs

```diff
@@ -128,8 +126,27 @@ impl Shell {
             Request::ProcessProposal(block) => {
                 tracing::debug!("Request ProcessProposal");
                 // TODO: use TM domain type in the handler
-                let (response, _tx_results) =
-                    self.process_proposal(block.into());
+                // NOTE: make sure to put any checks inside process_proposal
+                // since that function is called in other places to rerun the
+                // checks if (when) needed. Every check living outside that
+                // function will not be correctly replicated in the other
+                // locations
+                let (response, tx_results) =
+                    self.process_proposal(block.clone().into());
+                // Cache the response in case of future calls from Namada. If
+                // hash conversion fails avoid caching
+                if let Ok(block_hash) = block.hash.try_into() {
```

Comment:
copy/clone the block hash before to avoid cloning the entire block. that's the only field required from `block` on line 138

Reply:
Fixed in e2132f3c74d07884730e4a63a97b495022ae5228